### PR TITLE
Supports multiple ethereum API

### PR DESCRIPTION
**Pull request by clearloop**
## Desc

Migrates the new usage of primitives

### CODE REVIEW

File: src/mmr/runner.rs [hunk 1]
```diff
@@ -54,8 +51,8 @@ impl Runner {
             // Note:
             //
             // This trigger is ungly, need better solution in the future
-            if ptr % 30000 == 0 {
-                thread::spawn(move || Self::epoch(ptr as u64))
+            if (ptr + 15000) % 30000 == 0 {
+                thread::spawn(move || Self::epoch((ptr + 15000) as u64))
```
Comment (clearloop) [inline src/mmr/runner.rs:55]: Pre-run `epoch` before 15000 blocks

**Comment by clearloop** [comment]
https://github.com/darwinia-network/shadow/blob/bd7cbcfe1d4388426c440d9f49b07e8c93407988/src/api/ethereum/ffi.rs#L44

Got a problem parsing Ethereum API using FFI on the Golang side.

File: src/api/ethereum/ffi.rs [hunk 2]
```diff
@@ -10,16 +10,6 @@ struct GoString {
...
-impl From<String> for GoString {
-    fn from(s: String) -> GoString {
-        let c_tx = CString::new(s).expect("CString::new failed");
-        GoString {
-            a: c_tx.as_ptr(),
-            b: c_tx.as_bytes().len() as i64,
-        }
-    }
-}
-
```
Comment (clearloop) [inline src/api/ethereum/ffi.rs:22]: This rust pointer will be dropped by the garbage collection mechanism of golang

**Review by HackFisher — APPROVED** [review]

**Review by HackFisher — CHANGES_REQUESTED** [review]
Need further checking and testing

**Review by HackFisher — APPROVED** [review]

File: pkg/shadow/eth/receipt.go [hunk 3]
```diff
@@ -54,8 +54,8 @@ func GetChainBlockInfo(blockNum int64) (*BlockResult, error) {
 	}
 }
 
-func RPC(method string, params interface{}) (*dto.RequestResult, error) {
-	provider := providers.NewHTTPProvider(strings.ReplaceAll(API, "https://", ""), 10, true)
+func RPC(api string, method string, params interface{}) (*dto.RequestResult, error) {
+	provider := providers.NewHTTPProvider(strings.ReplaceAll(api, "https://", ""), 10, true)
 	pointer := &dto.RequestResult{}
```
Comment (HackFisher) [inline pkg/shadow/eth/receipt.go:59]: Could API be start with ```"http://```

**Pull request merged**